src/transformStateWithClones.js

```diff
@@ -1,13 +1,48 @@
 'use strict';
 
 /**
- * @param {Object} state
+ * @param {Object}  currentState
  * @param {Object[]} actions
  *
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear': {
+        currentState = {};
+        break;
+      }
+
+      case 'addProperties': {
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      }
+
+      case 'removeProperties': {
+        currentState = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          if (key in currentState) {
+            delete currentState[key];
+          }
+        }
+        break;
+      }
+
+      default:
+        // eslint-disable-next-line no-console
+        console.warn(`Неизвестный тип действия: ${action.type}`);
+        break;
+    }
+
+    states.push({ ...currentState });
```

Comment:
According to the checklist, it is better to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block, after the current action is processed. This line is correctly placed as per the checklist.